src/main/java/gregtech/common/metatileentities/multi/electric/MetaTileEntityCleanroom.java

```diff
@@ -0,0 +1,688 @@
+        if (lDist == 0) lDist = MIN_RADIUS;
+        if (rDist == 0) rDist = MIN_RADIUS;
+        if (bDist == 0) bDist = MIN_RADIUS;
+        if (fDist == 0) fDist = MIN_RADIUS;
+        if (hDist == 0) hDist = MIN_DEPTH;
+
+        // build each row of the structure
+        StringBuilder borderBuilder = new StringBuilder();     // BBBBB
+        StringBuilder wallBuilder = new StringBuilder();       // BXXXB
+        StringBuilder insideBuilder = new StringBuilder();     // X   X
+        StringBuilder roofBuilder = new StringBuilder();       // BFFFB
+        StringBuilder controllerBuilder = new StringBuilder(); // BFSFB
+
+        // everything to the left of the controller
+        for (int i = 0; i < lDist; i++) {
+            borderBuilder.append("B");
+            if (i == 0) {
+                wallBuilder.append("B");
+                insideBuilder.append("X");
+                roofBuilder.append("B");
+                controllerBuilder.append("B");
+            } else {
+                insideBuilder.append(" ");
+                wallBuilder.append("X");
+                roofBuilder.append("F");
+                controllerBuilder.append("F");
+            }
+        }
+
+        // everything in-line with the controller
+        borderBuilder.append("B");
+        wallBuilder.append("X");
+        insideBuilder.append(" ");
+        roofBuilder.append("F");
+        controllerBuilder.append("S");
+
+        // everything to the right of the controller
+        for (int i = 0; i < rDist; i++) {
+            borderBuilder.append("B");
+            if (i == rDist - 1) {
+                wallBuilder.append("B");
+                insideBuilder.append("X");
+                roofBuilder.append("B");
+                controllerBuilder.append("B");
+            } else {
+                insideBuilder.append(" ");
+                wallBuilder.append("X");
+                roofBuilder.append("F");
+                controllerBuilder.append("F");
+            }
+        }
+
+        // build each slice of the structure
+        String[] wall = new String[hDist + 1]; // "BBBBB", "BXXXB", "BXXXB", "BXXXB", "BBBBB"
+        Arrays.fill(wall, wallBuilder.toString());
+        wall[0] = borderBuilder.toString();
+        wall[wall.length - 1] = borderBuilder.toString();
+
+        String[] slice = new String[hDist + 1]; // "BBBBB", "X   X", "X   X", "X   X", "BFFFB"
+        Arrays.fill(slice, insideBuilder.toString());
+        slice[0] = borderBuilder.toString();
+        slice[slice.length - 1] = roofBuilder.toString();
+
+        String[] center = Arrays.copyOf(slice, slice.length); // "BBBBB", "X   X", "X   X", "X   X", "BFSFB"
+        center[center.length - 1] = controllerBuilder.toString();
+
+        TraceabilityPredicate wallPredicate = states(getCasingState(), getGlassState()).setMinGlobalLimited((((lDist + rDist + 1) * (bDist + fDist + 1) * (hDist + 1)) - ((lDist + rDist - 1) * (bDist + fDist - 1) * (hDist - 1))) * 3 / 4);
```

Comment:
The use of casing and glass states means that `borderBuilder` can be made of casings or glass, which means that you can have glass along the edges of the structure, which I am not sure is intended or not.